pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -220,15 +221,23 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 				// Remote side closed connection.
 				return receiveError
 			}
-			// This should be only set for the first request. Guard with ID check regardless.
-			if discReq.Node != nil && discReq.Node.Id != "" {
+			// This should be only set for the first request. Must be called even if the request has
+			// a missing node id (for example native gRPC)
+			if con.node == nil { // same condition is checked inside initConnection
 				if cancel, err := s.initConnection(discReq.Node, con); err != nil {
 					return err
 				} else if cancel != nil {
 					defer cancel()
 				}
 			}
 
+			// Based on node metadata a different generator was selected, use it instead of the default
+			// behaviour.
+			if con.node.Generator != nil && discReq.TypeUrl != EndpointType {
```

Comment:
Should the generator be per type, so we can have logic like `istio.io/virtualservice -> configgenerator, lds -> grpcgen, eds -> fallthrough` or similar? Not sure if its better or not. Just seems slightly weird to have endpoints special cased, but not a big deal

Reply:
Yes, that's the intent - still looking for best way to configure the per resource generators.

The interface is designed to allow per-resource generator, with fallback to per node and global.
The missing part is how to set them up - first step is to finish the bikeshed on the interface.